### ISSUE TITLE
Maximum call stack exceeded on import

### ISSUE THREAD

**Issue by JanMatas** [pyodide/pyodide#373]
Hi, 

I am trying to import numpy and matplotlib in service worker and I quite often receive an error:
```
Python initialization complete
pyodide.js:129 Loading matplotlib from default channel
pyodide.js:129 Loading pytz from default channel
pyodide.js:129 Loading python-dateutil from default channel
pyodide.js:129 Loading pyparsing from default channel
pyodide.js:129 Loading numpy from default channel
pyodide.js:129 Loading kiwisolver from default channel
pyodide.js:129 Loading cycler from default channel
Uncaught (in promise) RangeError: Maximum call stack size exceeded
    at wasm-function[1718]:0
    at wasm-function[1717]:9
    at wasm-function[1653]:18
    at wasm-function[12180]:7
    at wasm-function[1727]:214
    at wasm-function[2742]:44439
    at wasm-function[16565]:7
    at wasm-function[833]:245
    at wasm-function[836]:88
    at wasm-function[2747]:1127
```
Quite often means approximatelly 50% of the cases -> the problem is not deterministic. I am using `Google Chrome Version 73.0.3683.86 (Official Build) (64-bit)`and pyodide 0.10.0 (official build downloaded from the releases section of this repo.`

The service worker code is pretty simple:
```
self.languagePluginUrl = '/static/pyodide/'
importScripts('/static/pyodide/pyodide.js')

languagePluginLoader.then(() => {
  self.pyodide.runPythonAsync(`import numpy, matplotlib, matplotlib.pyplot`)
})
```

The worker is created like this:
```
      this.pyodideWorker = new Worker('/static/pyodide-worker.js')
```

Happy to provide better output if you release debug build or give me some advice how to build pyodide with debug symbols.  Also happy to help debugging/fixing and making PR, but I am currently at lost on where to start.

Any advice would be appreciated.

PS: I am aware that matplotlib requires DOM to work properly, so it would do nothing in webworker. I was planning to just monkey-patch it/create another backend which just returns base64 of the chart.

**Comment by mdboom** [comment]
The stack size thing is something that does show up occasionally in Chrome.  

I usually see this in code that creates infinite recursion in Python -- Python's own stack smashing check is supposed to kick-in and not let that become a C-level stack smash (which is what this is), and so there's a bit of code to decrease Python's stack limit so it never exceeds the browser's here: https://github.com/iodide-project/pyodide/blob/master/src/pyodide.js#L242

But I don't think that's what you're running into, since it's happening at import time.  It's possible that the stack size in a WebWorker is smaller than on the main thread (I have zero information to back that theory up, just throwing it out as something to investigate).

It's a little hard to say what might be happening since that stack trace doesn't have any function names.  In order to do that, you need to rebuild Pyodide with debugging symbols.  If you can successfully build Pyodide, then you can do:

```
EMCC_CFLAGS=--profiling make
```

to build a Pyodide with debugging symbols.  Then the stacktrace will include function names which might provide some clues as to the cause.  (Issue #366 is about doing this as part of our CI system so you could just download this rather than needing to build it yourself).

PS: matplotlib should work just fine as long as you don't use `show`, so have it use the `Agg` backend which will let you save figures of most of the file types.

**Comment by JanMatas** [comment]
Hi, thanks for the very fast response and detailed instructions.

I will report back here once I try it (should be by the end of week)!

**Comment by Aleksandar1932** [comment]
Reproduced in Chrome with pure python code without any dependencies, and works perfectly in Firefox. Has someone figured it out yet?

**Comment by hoodmane** [comment]
Well we know what's going on: Chrome has a much smaller stack than Firefox. In particular, the Firefox stack is approximately 3x larger. Currently Python stack frames hog a lot of stack space. See #1541.

I opened a bug report on chrome a long time ago complaining about the stack size and asking for an increase. Maybe some day they will do it?

**Comment by hoodmane** [comment]
Closed by #2019.
